src/drt/src/pa/FlexPA_prep.cpp

```diff
@@ -1842,100 +1842,102 @@ void FlexPA::genInstRowPattern(std::vector<frInst*>& insts)
     return;
   }
 
-  const int num_node = (insts.size() + 2) * ACCESS_PATTERN_END_ITERATION_NUM;
-
-  std::vector<FlexDPNode> nodes(num_node);
+  std::vector<std::vector<FlexDPNode>> nodes(insts.size() + 2);
 
   genInstRowPatternInit(nodes, insts);
   genInstRowPatternPerform(nodes, insts);
   genInstRowPattern_commit(nodes, insts);
 }
 
 // init dp node array for valid access patterns
-void FlexPA::genInstRowPatternInit(std::vector<FlexDPNode>& nodes,
+void FlexPA::genInstRowPatternInit(std::vector<std::vector<FlexDPNode>>& nodes,
                                    const std::vector<frInst*>& insts)
 {
   // init virtual nodes
-  const int start_node_idx
-      = getFlatIdx(-1, 0, ACCESS_PATTERN_END_ITERATION_NUM);
-  const int end_node_Idx
-      = getFlatIdx(insts.size(), 0, ACCESS_PATTERN_END_ITERATION_NUM);
-  nodes[start_node_idx].setNodeCost(0);
-  nodes[start_node_idx].setPathCost(0);
-  nodes[end_node_Idx].setNodeCost(0);
+  const int source_node_idx = insts.size() + 1;
+  nodes[source_node_idx] = std::vector<FlexDPNode>(1);
+  auto source_node = &(nodes[source_node_idx][0]);
```

Comment:
auto& source_node = nodes[source_node_idx][0]
use the object reference instead of pointer